bigquery_etl/shredder/delete.py

```diff
@@ -0,0 +1,337 @@
+    "--source_project",
+    help="override the project used for deletion request tables",
+)
+parser.add_argument(
+    "--target-project",
+    "--target_project",
+    help="override the project used for target tables",
+)
+parser.add_argument(
+    "--parallelism",
+    default=4,
+    type=int,
+    help="Maximum number of queries to execute concurrently",
+)
+parser.add_argument(
+    "--max-single-dml-bytes",
+    "--max_single_dml_bytes",
+    default=10 * 2 ** 40,
+    type=int,
+    help="Maximum number of bytes in a table that should be processed using a single "
+    "DML query; tables above this limit will be processed using per-partition "
+    "queries; this option prevents queries against large tables from exceeding the "
+    "6-hour time limit; defaults to 10 TiB",
+)
+parser.add_argument(
+    "--end-date",
+    "--end_date",
+    default=datetime.utcnow().date(),
+    type=lambda x: datetime.strptime(x, "%Y-%m-%d").date(),
+    help="last date of deletion requests to process and last date of pings to delete, "
+    "defaults to today in UTC",
+)
+parser.add_argument(
+    "--start-date",
+    "--start_date",
+    type=lambda x: datetime.strptime(x, "%Y-%m-%d").date(),
+    help="first date of deletion requests to process; DOES NOT apply to ping date; "
+    "defaults to 14 days before --end-date in UTC",
+)
+parser.add_argument(
+    "--priority",
+    default=bigquery.QueryPriority.INTERACTIVE,
+    type=str.upper,
+    choices=[bigquery.QueryPriority.BATCH, bigquery.QueryPriority.INTERACTIVE],
+    help="Priority for BigQuery query jobs; BATCH priority may significantly slow "
+    "down queries if reserved slots are not enabled for the billing project; "
+    "INTERACTIVE priority is limited to 100 concurrent queries per project",
+)
+parser.add_argument(
+    "--billing_projects",
+    "--billing-projects",
+    "--billing_project",
+    "--billing-project",
+    nargs="+",
+    default=["moz-fx-data-bq-batch-prod"],
+    help="One or more billing projects over which bigquery jobs should be distributed; "
+    "if not specified use the bigquery-batch-prod project",
+)
+add_table_filter_arguments(parser)
+
+WHERE_CLAUSE = """
+WHERE
+  {target.id_} IN (
+    SELECT DISTINCT
+      {source.id_}
+    FROM
+      `{source.sql_table_id}`
+    WHERE
+      {source_condition}
+  )
+  AND {partition_condition}
+"""
+
+DELETE_TEMPLATE = f"""
+DELETE
+  `{{target.sql_table_id}}`
+{WHERE_CLAUSE.strip()}
+"""
+
+SELECT_TEMPLATE = f"""
+CREATE TABLE
+  `{{destination.project}}.{{destination.dataset_id}}.{{destination.table_id}}`
+PARTITION BY
+  {{partition_expr}}
+{{clustering}}
+OPTIONS (
+  expiration_timestamp = '{{expiration_timestamp}}'
+)
+AS
+SELECT
+  *
+FROM
+  `{{target.sql_table_id}}`
+{WHERE_CLAUSE.strip()}
+  AND {{cluster_condition}}
+"""
+
+
+async def gather_all(futures):
+    """Gather results, raise last exception, and log other exceptions."""
+    results = await asyncio.gather(*futures, return_exceptions=True)
```

Comment:
Note that I am only passingly familiar with asyncio in python, so not going to be able to provide much validation of the task handling here.